In [1]:
import sqlalchemy
from sqlalchemy import create_engine
import pandas as pd
import io
%load_ext sql

Deploy AI and data apps for free on Ploomber Cloud! Learn more: https://docs.cloud.ploomber.io/en/latest/quickstart/signup.html


In [2]:
%sql postgresql://postgres:4901@localhost/airbnb
engine = create_engine('postgresql://postgres:4901@localhost/airbnb')

Connecting to 'postgresql://postgres:***@localhost/airbnb'

In [3]:
# %%sql
# SELECT *
# FROM listings
# LIMIT 1;

Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-------+------+------------------------------------------------------------+---------+-----------+---------------------+---------------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+---------+
| index |  id  |                            name                            | host_id | host_name | neighbourhood_group | neighbourhood | latitude | longitude |    room_type    | price | minimum_nights | number_of_reviews | last_review | reviews_per_month | calculated_host_listings_count | availability_365 | number_of_reviews_ltm | license |
+-------+------+------------------------------------------------------------+---------+-----------+---------------------+---------------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+---------+
|   0   | 5456 | Guesthouse in Austin · ★4.84 · 1 bedroom · 2 beds · 1 bath |   8028  |   Sylvia  |         None        |     78702     | 30.26057 | -97.73441 | Entire home/apt |  126  |       2        |        657        |  2023-08-27 |        3.72       |               1                |       306        |           42          |   None  |
+-------+------+------------------------------------------------------------+---------+-----------+---------------------+---------------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+---------+

## Pg Admin 4 Tasks
### Change Headers Names
Some columns require some better descriptor or column names to properly describe the kind of data in that column.
1. neighourhood to be changed to zip code.
2. room_type to be changed to rental.

### Drop unnecessary columns
1. Drop license column.

In [3]:
%%sql
SELECT *
FROM listings
# WHERE name LIKE '%Austin%'
LIMIT 1;

Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-------+------+------------------------------------------------------------+---------+-----------+---------------------+----------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+
| index |  id  |                            name                            | host_id | host_name | neighbourhood_group | zip_code | latitude | longitude |      rental     | price | minimum_nights | number_of_reviews | last_review | reviews_per_month | calculated_host_listings_count | availability_365 | number_of_reviews_ltm |
+-------+------+------------------------------------------------------------+---------+-----------+---------------------+----------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+
|   0   | 5456 | Guesthouse in Austin · ★4.84 · 1 bedroom · 2 beds · 1 bath |   8028  |   Sylvia  |         None        |  78702   | 30.26057 | -97.73441 | Entire home/apt |  126  |       2        |        657        |  2023-08-27 |        3.72       |               1                |       306        |           42          |
+-------+------+------------------------------------------------------------+---------+-----------+---------------------+----------+----------+-----------+-----------------+-------+----------------+-------------------+-------------+-------------------+--------------------------------+------------------+-----------------------+

### Introduction
Before going into the analytics of this project, it is important to understand our dataset and the kind of insights we can draw from it.
First, we need the number of listings we have, making sure there are no duplicate listings.

In [5]:
%%sql
SELECT COUNT(id)
FROM listings;


Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-------+
| count |
+-------+
| 14861 |
+-------+

In [6]:
%%sql
SELECT DISTINCT COUNT(id)
FROM listings;

Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-------+
| count |
+-------+
| 14861 |
+-------+

From our dataset column calculated_host_listings, we know there are some hosts with more than one listing. Let find the number of unique hosts we have for the 14861 distinct listings.

In [8]:
%%sql
SELECT COUNT(DISTINCT host_id)
FROM listings
WHERE id IS NOT NULL;

Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-------+
| count |
+-------+
|  8834 |
+-------+

There are 8834 unique hosts, who have at least one listing. Next is to find out how many rentals we have and the number of units per every rental.

In [11]:
%%sql
SELECT rental, COUNT(id) AS unit_count
FROM listings
GROUP BY rental
ORDER BY unit_count DESC;

Running query in 'postgresql://postgres:***@localhost/airbnb'

4 rows affected.

+-----------------+------------+
|      rental     | unit_count |
+-----------------+------------+
| Entire home/apt |   12205    |
|   Private room  |    2529    |
|   Shared room   |    121     |
|    Hotel room   |     6      |
+-----------------+------------+

Adding up all the units of the different rentals, we get 14861 units, which validates the answer and some previous queries.

Now, we need to go a bit deeper and find out the distribution of the units among the hosts. Then, we need to find the host(s) with the most listings in every rental.

In [20]:
%config SqlMagic.displaylimit = None

displaylimit: Value None will be treated as 0 (no limit)

In [25]:
%%sql
SELECT 
CASE
WHEN calculated_host_listings_count = 1 THEN '1 Unit'
WHEN calculated_host_listings_count = 2 THEN '2 Units'
WHEN calculated_host_listings_count = 3 THEN '3 Units'
WHEN calculated_host_listings_count = 4 THEN '4 Units'
WHEN calculated_host_listings_count >= 5 AND calculated_host_listings_count <= 9 THEN '5+ Units'
WHEN calculated_host_listings_count >= 10 THEN '10+ Units' END AS units_distribution, rental, COUNT(host_id) AS host_id_count
FROM listings
GROUP BY rental, units_distribution
ORDER BY host_id_count DESC
LIMIT 5; 

Running query in 'postgresql://postgres:***@localhost/airbnb'

5 rows affected.

+--------------------+-----------------+---------------+
| units_distribution |      rental     | host_id_count |
+--------------------+-----------------+---------------+
|       1 Unit       | Entire home/apt |      5880     |
|     10+ Units      | Entire home/apt |      2740     |
|      2 Units       | Entire home/apt |      1507     |
|       1 Unit       |   Private room  |      1176     |
|      5+ Units      | Entire home/apt |      994      |
+--------------------+-----------------+---------------+

Killing two birds with one stone, we know the distribution of the rentals among the host. The highest unit distribution is the 1 Unit distribution to 1 host and 1 hotel room to 1 host is the lowest unit distribution. This query also shows that a very high number of hosts have over 10 units.
That's interesting. Let's find the top 5 hosts with the most listings.

In [30]:
%%sql
SELECT host_id,host_name,COUNT(id) AS listings_count
FROM listings
GROUP BY host_id, host_name
ORDER BY listings_count DESC
LIMIT 5;

Running query in 'postgresql://postgres:***@localhost/airbnb'

5 rows affected.

+-----------+------------+----------------+
|  host_id  | host_name  | listings_count |
+-----------+------------+----------------+
| 107434423 | Blueground |      136       |
| 274552333 |  Evonify   |      126       |
| 263502162 |  Landing   |      112       |
|  8167447  |   Martin   |       89       |
| 501999514 | RoomPicks  |       85       |
+-----------+------------+----------------+

From the query, Blueground has over 136 units. It would be great to see the distribution of these units.

In [32]:
%%sql
SELECT rental, COUNT(id) AS listings_count
FROM listings
WHERE host_id = 107434423
GROUP BY rental
ORDER BY listings_count;

Running query in 'postgresql://postgres:***@localhost/airbnb'

1 rows affected.

+-----------------+----------------+
|      rental     | listings_count |
+-----------------+----------------+
| Entire home/apt |      136       |
+-----------------+----------------+

Blueground has 136 full apartments. Interesting. A quick google search shows that Blueground is a company, which makes more sense. https://www.theblueground.com/ Martin on the other hand has probably a few apartments in the 10s or twenties but has every single room listed.

In [33]:
%%sql
SELECT rental, COUNT(id) AS listings_count
FROM listings
WHERE host_id = 8167447 
GROUP BY rental
ORDER BY listings_count;

Running query in 'postgresql://postgres:***@localhost/airbnb'

2 rows affected.

+--------------+----------------+
|    rental    | listings_count |
+--------------+----------------+
| Private room |       33       |
| Shared room  |       56       |
+--------------+----------------+

Next, we need to find the availability of the unit rentals in months. First ensure we do not have any negatives or numbers over 366 days in a leap year.

In [35]:
%%sql
SELECT availability_365
FROM listings
WHERE availability_365 > 366 OR availability_365 < 0;

Running query in 'postgresql://postgres:***@localhost/airbnb'

+------------------+
| availability_365 |
+------------------+
+------------------+

In [39]:
%%sql
SELECT
CASE
WHEN availability_365 < 30 THEN 'Less than 1 month'
WHEN availability_365 < 60 THEN 'Less than 2 months'
WHEN availability_365 < 90 THEN 'Less than 3 months'
WHEN availability_365 < 120 THEN 'Less than a quarter'
WHEN availability_365 < 150 THEN 'Less than 5 months'
WHEN availability_365 < 180 THEN 'Less than half a year'
WHEN availability_365 < 299 THEN '6 - 9 months'
WHEN availability_365 < 300 THEN 'Almost all year'
WHEN availability_365 < 365 THEN 'All year' END AS availability, COUNT(id) AS listings_count, COUNT(host_id) AS host_count
FROM listings
GROUP BY availability
ORDER BY listings_count DESC, host_count DESC;


Running query in 'postgresql://postgres:***@localhost/airbnb'

10 rows affected.

+-----------------------+----------------+------------+
|      availability     | listings_count | host_count |
+-----------------------+----------------+------------+
|   Less than 1 month   |      4546      |    4546    |
|        All year       |      3485      |    3485    |
|      6 - 9 months     |      1974      |    1974    |
|   Less than 3 months  |      1277      |    1277    |
| Less than half a year |      1045      |    1045    |
|   Less than 2 months  |      875       |    875     |
|   Less than 5 months  |      776       |    776     |
|  Less than a quarter  |      588       |    588     |
|          None         |      277       |    277     |
|    Almost all year    |       18       |     18     |
+-----------------------+----------------+------------+

It seems every host sticks with a particular availability all year round.